### ISSUE TITLE
Support OSCP certificate revocation with Let's Encrypt

### ISSUE THREAD

**Issue by ddtmachado** [traefik/traefik#9615]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

Traefik provides automatic integration for generating certificates with Let's Encrypt, using the ACME protocol, but currently no validation is made to check if the generated certificate is still valid and not revoked on the CA.

Let's Encrypt on the other hand supports the [OCSP protocol](https://www.rfc-editor.org/rfc/rfc2560) and [provides an endpoint](https://letsencrypt.org/docs/lencr.org/) to check for the certificate status with it. The information about the OCSP endpoint is also included in the certificate itself as an extension field, as defined on the RFC.

### Implementation Proposal

Since Traefik's [TLS Manager](https://github.com/traefik/traefik/blob/master/pkg/tls/tlsmanager.go) can already make the distinction of ACME certificates, it would be a matter of extracting the OCSP endpoint and create a new background task to fetch the certificate status data periodically. The frequency could be a new configurable item under the certificate resolver object if adjusting it proves beneficial in any use case.

According to the certificate status response the following actions would be taken:

- **good**
    nothing to do
- **revoked**
    remove the certificate from the store
    trigger a new certificate request for the domain
    To be defined: disable the router or allow it to continue with the default certificate?
- **unknown**
    nothing to do (soft fail)

**Comment by sh7dm** [comment]
Maybe this could be done for user-supplied certificates as well, because they can also be revoked and forgotten. Having some status about that would be helpful.

**Comment by pkern** [comment]
I'm still somewhat unconvinced of this. If a certificate has been revoked, there is usually a reason to it. If you keep renewing behind the scenes without addressing the root cause you might not be very successful at keeping a non-revoked certificate (or still being able to request one via ACME - as the key material might have been blocked or your account might get rate-limited).

At the very least, not only do you need to create a new certificate request but you also need to regenerate key material when requesting a new cert - especially when the revocation reason is keyCompromise. Given some misunderstandings in the OCSP stapling bug I feel obliged to point this out.

https://wiki.mozilla.org/CA/Revocation_Reasons documents the various different reasons a CA can present for a certificate being revoked.

**Comment by ddtmachado** [comment]
Hi @pkern , I understand your concern but dealing with the root cause of a certificate revocation is out of scope in the OCSP context, if not for Traefik at all. 

We can check the status and then react by removing the certificate from the store so it is not in use anymore. That by itself would trigger a new certificate request, since the TLS enabled routers would still need one. 
At that point we can only hope the CA would have dealt with the issue if it was a key compromise.

**Comment by pkern** [comment]
At least I hope that'd clear key material? Doesn't solve the rate limit concern. In the worst case the system won't let you renew by the time you remediated the issue. But maybe Traefik already rate limits itself.

**Comment by ddtmachado** [comment]
Indeed, the [certificatesDuration](https://doc.traefik.io/traefik/https/acme/#certificatesduration) option provides safe defaults that helps any Traefik instance to stay out of the rate limit from LE in most cases.
